### PR TITLE
Optimize the use of GitHub Actions resources

### PR DESCRIPTION
**Pull request by J535D165**
This PR proposes to optimize the GitHub Actions workflows. 

- Don't run linters in the matrix. Ruff and other linters are (in general) Python version-independent and platform-independent. Use linux for linters. 
- Run docs test outside of the matrix and run against Python version of .readthedocs.yaml.
- Split test dependencies into test and docs. Docs deps are more expensive to install. 
- Make the matrix fail-fast. This library is unlikely to have many issues with the various Python versions, so fail-fast is most likely good enough. 
- Don't run integration tests on documentation and tutorial updates. 

Suggestions: 
- remove Pylint. It's slow and most functionality is covered nowadays by ruff. 
- Use ruff format for formatting

### CODE REVIEW

**Comment by qubixes** [comment]
Hi @J535D165 thanks for the PR! One thing that I am a little worried about, is not running the linters/ruff/mypy on the whole matrix. I think in general for packages that you are right that most Python/OS specific issues should be covered by the unit/integration tests. However, in our case we have only a tiny fraction (less than 5% of the codebase) tested in the matrix, while the integration tests that do cover most of the code are only run on a single OS/Python version. Obviously, this is not the idea setup, but I'm worried that some obvious things that could be caught by the linters (most frequent example, using Python<3.10 style type hints, without the `__future__` import) can now slip past our CI tests. What do you think?

**Comment by J535D165** [comment]
I don't think it's an issue. I tested your example for demo purposes, and it fails nicely (see above).

**Comment by qubixes** [comment]
> I don't think it's an issue. I tested your example for demo purposes, and it fails nicely (see above).

In most cases, yes the tests will fail, but I think there are still some edge cases that could pass through that ruff might catch. What about just adding ruff to the matrix? It doesn't take much extra time anyway, and it might catch something every now and then for older versions of Python.

**Comment by J535D165** [comment]
I disagree. See also docs for target-verion. 

Maybe it will be interesting to improve the coverage soon.

**Comment by qubixes** [comment]
I don't think there is that much low hanging fruit in terms of improving coverage (the numbers won't be super stellar, since a lot of the iBridges code-base deals with error-catching and those depend also on the server configuration). But those tests are not in the matrix, so are only tested for a single Python/OS version. Perhaps in the future we can spin up multiple test containers for a single server, but I personally don't know how that would or should work.

File: .github/workflows/integration-tests-irods.yml [hunk 1]
```diff
@@ -4,7 +4,13 @@ on:
   push:
     branches:
       - develop
+    paths-ignore:
+      - 'docs/**'
+      - 'tutorials/**'
```
Comment (chStaiger) [inline .github/workflows/integration-tests-irods.yml:9]:
```suggestion
      - 'tutorials/**'
      - 'README.md'
```

File: .github/workflows/integration-tests-irods.yml [hunk 2]
```diff
@@ -4,7 +4,13 @@ on:
...
   pull_request:
+    paths-ignore:
+      - 'docs/**'
+      - 'tutorials/**'
```
Comment (chStaiger) [inline .github/workflows/integration-tests-irods.yml:13]:
```suggestion
      - 'tutorials/**'
      - 'README.md'
```

File: .github/workflows/integration-tests-yoda.yml [hunk 3]
```diff
@@ -4,7 +4,13 @@ on:
   push:
     branches:
       - develop
+    paths-ignore:
+      - 'docs/**'
+      - 'tutorials/**'
```
Comment (chStaiger) [inline .github/workflows/integration-tests-yoda.yml:9]:
```suggestion
      - 'tutorials/**'
      - 'README.md'
```

File: .github/workflows/integration-tests-yoda.yml [hunk 4]
```diff
@@ -4,7 +4,13 @@ on:
...
   pull_request:
+    paths-ignore:
+      - 'docs/**'
+      - 'tutorials/**'
```
Comment (chStaiger) [inline .github/workflows/integration-tests-yoda.yml:13]:
```suggestion
      - 'tutorials/**'
      - 'README.md'
```

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -15,47 +15,75 @@ on:
 jobs:
```
Comment (chStaiger) [inline .github/workflows/main.yml:15]: Can you please add the paths-ignore for the python tutorials and README here too?

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -15,47 +15,75 @@ on:
...
     strategy:
-      fail-fast: false
+      fail-fast: true
       matrix:
         os: [ubuntu-latest]
         python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
         include:
           - os: windows-latest
-            python-version: "3.11"
+            python-version: "3.12"
```
Comment (chStaiger) [inline .github/workflows/main.yml:24]: Good spot! Thank you!

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -15,47 +15,75 @@ on:
...
+        run: |
+          pip install --upgrade pip
+          pip install ".[test]"
 
+      - name: Test with pytest
+        run: |
+          pytest tests
+
+  linting:
+    runs-on: ubuntu-latest
```
Comment (chStaiger) [inline .github/workflows/main.yml:49]: I agree with @qubixes. We want to offer iBridges for all three operating systems an in the past the linting has pointed us to some incompatibilities. So I would keep it in. When/if we extend the coverage of the integration tests we can reduce the linting to linux.

**Review by chStaiger — CHANGES_REQUESTED** [review]
Many thanks for the many useful suggestions!!! Especially skipping integration tests when just updating the docs will help very much reducing our carbon footprint!

I am not so sure about the RTD documentation. You suggest to only test the rendering for python 3.9. Could that lead to mal-rendered docs on the RTD server. I am not sure which python version they use and if we arethen still compatible with them.

File: .github/workflows/main.yml (hunk 5 above)
Comment (J535D165) [reply .github/workflows/main.yml:15]: Ruff also lints the Jupyter notebooks (https://docs.astral.sh/ruff/configuration/#jupyter-notebook-discovery). Excluding the tutorials is not advised. I'm not sure if the README impacts the document rendering, so I left it out.

Comment (chStaiger) [reply .github/workflows/main.yml:15]: Ah, ok.

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -15,47 +15,75 @@ on:
...
+        uses: actions/checkout@v4
+
+      - name: Install dependencies
+        run: |
+          pip install --upgrade pip
+          pip install ".[test]"
 
+      - name: Test with pytest
+        run: |
+          pytest tests
```
Comment (chStaiger) [inline .github/workflows/main.yml:46]:
```suggestion
      - name: Test with pytest
        run: |
          pytest tests
          
      - name: Lint with PyLint
        run: |
          pylint ibridges
          
      - name: Check types with MyPy
        run: |
          mypy ibridges
```

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -15,47 +15,75 @@ on:
...
+        run: |
+          ruff check ibridges
+
+      - name: Lint with PyLint
+        run: |
+          pylint ibridges
+
+      - name: Check types with MyPy
+        run: |
+          mypy ibridges
```
Comment (chStaiger) [inline .github/workflows/main.yml:71]:
```suggestion

```

File: .github/workflows/main.yml (hunk 7 above)
Comment (chStaiger) [reply .github/workflows/main.yml:49]: I moved the pylint and mypy up to the orginial place again. 

In theory you are right it should be enough to test for lunix should be enough. However,  in the past development cycles we encountered some descrepancies between the linting/mypy results, most probably due to that we do not fully understand the whole stack from OS, through PRC to the iRODS server. Nevertheless, the linting in this section of the workflow prevented us from releasing buggy software for MAc and Windows. So I would like to keep them here.

**Comment by chStaiger** [comment]
@J535D165 I cannot accept the latest suggestions I made on your branch. Can you please accept them? I would like to merge your changes into develop.
